File: src/orchestwin/artifacts/design_fixtures.py
"""Shared deterministic fixtures for Sprint 06 design artifact tests."""

from __future__ import annotations

from datetime import UTC, datetime
from uuid import UUID

from orchestwin.artifacts.design import (
    DesignApproach,
    create_design_alternative,
    create_design_workflow,
    create_synthetic_design_critique,
)
from orchestwin.artifacts.prototypes import (
    PrototypeElementKind,
    PrototypeScreenState,
    PrototypeViewport,
    create_declarative_prototype,
    create_prototype_element,
    create_prototype_screen,
    create_prototype_transition,
)
from orchestwin.projects.requirements import (
    RequirementKind,
    RequirementPriority,
    create_requirement,
    create_user_story,
)
from orchestwin.projects.requirements_primitives import (
    RequirementsContextKind,
    RequirementsContextReference,
    RequirementSourceKind,
    RequirementSourceReference,
    UserTwinVersionReference,
)
from orchestwin.projects.requirements_quality import (
    DefinitionOfDoneApplicability,
    VerificationMethod,
    create_acceptance_criterion,
    create_definition_of_done_item,
    create_usage_scenario,
)
from orchestwin.projects.requirements_specifications import (
    RequirementsSpecificationVersion,
    create_requirements_specification,
)
from orchestwin.twins.epistemics import (
    ConfidenceScore,
    EvidenceReference,
    EvidenceSourceKind,
    ObservationProvenance,
)
from test.python.artifacts.design_packages import (
    DesignPackageVersion,
    create_design_concern,
    create_design_exploration_package,
    create_design_grounding,
)

PROJECT_ID = UUID("00000000-0000-4000-8000-000000000001")
OWNER_ID = UUID("00000000-0000-4000-8000-000000000002")
REQUIREMENTS_VERSION_ID = UUID("00000000-0000-4000-8000-000000000003")
REQUIREMENT_ID = UUID("00000000-0000-4000-8000-000000000010")
STORY_ID = UUID("00000000-0000-4000-8000-000000000020")
CRITERION_ID = UUID("00000000-0000-4000-8000-000000000030")
SCENARIO_ID = UUID("00000000-0000-4000-8000-000000000040")
DOD_ID = UUID("00000000-0000-4000-8000-000000000050")
TWIN_ID = UUID("00000000-0000-4000-8000-000000000060")
ALTERNATIVE_ONE_ID = UUID("00000000-0000-4000-8000-000000000070")
ALTERNATIVE_TWO_ID = UUID("00000000-0000-4000-8000-000000000071")
PROTOTYPE_ID = UUID("00000000-0000-4000-8000-000000000080")
DESIGN_VERSION_ID = UUID("00000000-0000-4000-8000-000000000090")
CREATED_AT = datetime(2026, 8, 19, 9, 0, tzinfo=UTC)


def requirement_context(
    kind: RequirementsContextKind,
    ordinal: int,
) -> RequirementsContextReference:
    """Create one exact Requirements-stage context reference."""
    return RequirementsContextReference(
        kind=kind,
        artifact_id=UUID(int=ordinal),
        version_number=1,
        content_hash=f"{ordinal:x}" * 64,
    )


def twin_reference() -> UserTwinVersionReference:
    """Create the exact User Twin used throughout design tests."""
    return UserTwinVersionReference(
        twin_id=TWIN_ID,
        version_number=2,
        content_hash="a" * 64,
        name="Hotel Receptionist Twin",
    )


def requirements_version() -> RequirementsSpecificationVersion:
    """Create one complete immutable Requirements baseline."""
    source = RequirementSourceReference(
        kind=RequirementSourceKind.PROJECT_BRIEF,
        source_id="brief-version",
        source_version=1,
        content_hash="b" * 64,
        locator="functional_requirements[0]",
    )
    requirement = create_requirement(
        requirement_id=REQUIREMENT_ID,
        code="REQ-001",
        title="Create reservations",
        statement="The system must create reservations.",
        kind=RequirementKind.FUNCTIONAL,
        priority=RequirementPriority.MUST,
        sources=(source,),
        user_twin_references=(twin_reference(),),
    )
    story = create_user_story(
        story_id=STORY_ID,
        code="USR-001",
        user_twin_reference=twin_reference(),
        goal="create a reservation",
        benefit="serve a guest accurately",
        requirement_ids=(REQUIREMENT_ID,),
    )
    criterion = create_acceptance_criterion(
        criterion_id=CRITERION_ID,
        code="AC-001",
        statement="A reservation receives a unique identifier.",
        verification_method=VerificationMethod.AUTOMATED_TEST,
        requirement_ids=(REQUIREMENT_ID,),
        user_story_ids=(STORY_ID,),
    )
    scenario = create_usage_scenario(
        scenario_id=SCENARIO_ID,
        code="SCN-001",
        title="Create a reservation",
        actor=twin_reference(),
        preconditions=(),
        trigger="A guest requests a room.",
        steps=("Save the reservation.",),
        expected_outcome="The reservation can be retrieved.",
        requirement_ids=(REQUIREMENT_ID,),
        acceptance_criterion_ids=(CRITERION_ID,),
    )
    done = create_definition_of_done_item(
        item_id=DOD_ID,
        code="DOD-001",
        statement="All automated acceptance tests pass.",
        verification_method=VerificationMethod.AUTOMATED_TEST,
        applicability=DefinitionOfDoneApplicability.REQUIRED,
        requirement_ids=(REQUIREMENT_ID,),
    )
    specification = create_requirements_specification(
        project_id=PROJECT_ID,
        project_brief_reference=requirement_context(
            RequirementsContextKind.PROJECT_BRIEF,
            11,
        ),
        agent_team_reference=requirement_context(
            RequirementsContextKind.AGENT_TEAM,
            12,
        ),
        user_modeling_reference=requirement_context(
            RequirementsContextKind.USER_MODELING,
            13,
        ),
        catalog_version=1,
        catalog_content_hash="c" * 64,
        user_twin_references=(twin_reference(),),
        requirements=(requirement,),
        user_stories=(story,),
        acceptance_criteria=(criterion,),
        scenarios=(scenario,),
        risks=(),
        definition_of_done=(done,),
    )

    return RequirementsSpecificationVersion(
        id=REQUIREMENTS_VERSION_ID,
        project_id=PROJECT_ID,
        version_number=1,
        specification=specification,
        content_hash=specification.content_hash,
        created_by_user_id=OWNER_ID,
        created_at=CREATED_AT,
    )


def design_alternative(
    *,
    index: int,
):
    """Create one of two deliberately different design alternatives."""
    alternative_id = ALTERNATIVE_ONE_ID if index == 1 else ALTERNATIVE_TWO_ID
    approach = DesignApproach.GUIDED_WORKFLOW if index == 1 else DesignApproach.DASHBOARD_FIRST
    workflow = create_design_workflow(
        workflow_id=UUID(int=100 + index),
        code=f"FLOW-{index:03d}",
        title="Create a reservation",
        steps=("Review availability.", "Save the reservation."),
        requirement_ids=(REQUIREMENT_ID,),
        user_story_ids=(STORY_ID,),
    )

    return create_design_alternative(
        alternative_id=alternative_id,
        code=f"DES-{index:03d}",
        approach=approach,
        title=("Guided reservation flow" if index == 1 else "Reservation operations dashboard"),
        summary=(
            "Guide the receptionist through one decision at a time."
            if index == 1
            else "Keep availability and reservation actions visible together."
        ),
        rationale=(
            "Reduce cognitive load for occasional users."
            if index == 1
            else "Reduce navigation for experienced users."
        ),
        requirement_ids=(REQUIREMENT_ID,),
        user_story_ids=(STORY_ID,),
        acceptance_criterion_ids=(CRITERION_ID,),
        user_twin_references=(twin_reference(),),
        workflows=(workflow,),
        information_architecture=("Availability", "Reservation", "Confirmation"),
        accessibility_considerations=("All controls have persistent labels",),
        security_considerations=("Guest data is minimized in summaries",),
        advantages=("Clear progression" if index == 1 else "Fast access to frequent actions",),
        trade_offs=("More navigation" if index == 1 else "Higher information density",),
    )


def critique(
    *,
    index: int,
):
    """Create one synthetic critique for the selected alternative index."""
    alternative_id = ALTERNATIVE_ONE_ID if index == 1 else ALTERNATIVE_TWO_ID
    provenance = ObservationProvenance.from_references(
        (
            EvidenceReference(
                source_kind=EvidenceSourceKind.MODEL_OUTPUT,
                source_id="fake-design-provider:1",
                source_version=1,
                content_hash="d" * 64,
                locator=f"alternatives.DES-{index:03d}",
                summary="Deterministic synthetic User Twin critique.",
            ),
        )
    )

    return create_synthetic_design_critique(
        critique_id=UUID(int=200 + index),
        code=f"CRQ-{index:03d}",
        design_alternative_id=alternative_id,
        user_twin_reference=twin_reference(),
        strengths=("The main task is visible.",),
        concerns=("The flow may require adaptation for different experience levels.",),
        suggested_changes=("Preserve visible recovery actions.",),
        provenance=provenance,
        confidence=ConfidenceScore(0.65),
        rationale="The critique is inferred from the approved User Twin profile.",
    )


def prototype():
    """Create a prototype for the owner-selected first alternative."""
    input_element = create_prototype_element(
        element_id=UUID(int=301),
        code="ELM-001",
        kind=PrototypeElementKind.TEXT_INPUT,
        content="Guest name",
        accessible_name="Guest name",
        requirement_ids=(REQUIREMENT_ID,),
        field_name="guest_name",
        required=True,
    )
    button = create_prototype_element(
        element_id=UUID(int=302),
        code="ELM-002",
        kind=PrototypeElementKind.BUTTON,
        content="Save reservation",
        accessible_name="Save reservation",
        acceptance_criterion_ids=(CRITERION_ID,),
    )
    status = create_prototype_element(
        element_id=UUID(int=303),
        code="ELM-003",
        kind=PrototypeElementKind.STATUS,
        content="Reservation saved",
    )
    entry = create_prototype_screen(
        screen_id=UUID(int=310),
        code="SCR-001",
        title="Create reservation",
        state=PrototypeScreenState.DEFAULT,
        elements=(input_element, button),
        requirement_ids=(REQUIREMENT_ID,),
        user_story_ids=(STORY_ID,),
    )
    confirmation = create_prototype_screen(
        screen_id=UUID(int=311),
        code="SCR-002",
        title="Reservation confirmation",
        state=PrototypeScreenState.SUCCESS,
        elements=(status,),
        acceptance_criterion_ids=(CRITERION_ID,),
    )
    transition = create_prototype_transition(
        transition_id=UUID(int=320),
        code="TRN-001",
        source_screen_id=entry.id,
        trigger_element_id=button.id,
        target_screen_id=confirmation.id,
        outcome="The confirmation screen becomes visible.",
    )

    return create_declarative_prototype(
        prototype_id=PROTOTYPE_ID,
        code="PRT-001",
        title="Reservation flow prototype",
        design_alternative_id=ALTERNATIVE_ONE_ID,
        entry_screen_id=entry.id,
        screens=(entry, confirmation),
        transitions=(transition,),
        supported_viewports=(PrototypeViewport.MOBILE, PrototypeViewport.DESKTOP),
    )


def design_package(
    *,
    selected: bool = True,
    include_prototype: bool = True,
):
    """Create one complete Design Package fixture."""
    requirements = requirements_version()
    owner_selection = ALTERNATIVE_ONE_ID if selected else None
    package_prototype = prototype() if selected and include_prototype else None

    return create_design_exploration_package(
        project_id=PROJECT_ID,
        grounding=create_design_grounding(requirements),
        alternatives=(design_alternative(index=2), design_alternative(index=1)),
        critiques=(critique(index=2), critique(index=1)),
        recommended_alternative_id=ALTERNATIVE_ONE_ID,
        owner_selected_alternative_id=owner_selection,
        prototype=package_prototype,
        concerns=(
            create_design_concern(
                concern_id=UUID(int=400),
                code="DRK-001",
                summary="Expert users may find the guided flow slower.",
                mitigation="Keep keyboard shortcuts visible in later revisions.",
                requirement_ids=(REQUIREMENT_ID,),
                design_alternative_ids=(ALTERNATIVE_ONE_ID,),
            ),
        ),
        open_questions=("Should expert mode be included in the first release?",),
    )


def design_version(
    *,
    version_number: int = 1,
    package=None,
) -> DesignPackageVersion:
    """Create one immutable Design Package version."""
    resolved_package = package if package is not None else design_package()

    return DesignPackageVersion(
        id=DESIGN_VERSION_ID if version_number == 1 else UUID(int=500 + version_number),
        project_id=PROJECT_ID,
        version_number=version_number,
        based_on_version_number=None if version_number == 1 else version_number - 1,
        package=resolved_package,
        content_hash=resolved_package.content_hash,
        created_by_user_id=OWNER_ID,
        created_at=CREATED_AT,
    )

File: frontend/src/test/designFixtures.ts
import { buildSelectedDesignPackage } from "../components/designPrototype";
import type {
  DesignPackageDiffPayload,
  DesignPackagePayload,
  DesignPackageVersionPayload,
  HumanGatePayload,
} from "../types/design";

export const DESIGN_PROJECT_ID = "00000000-0000-4000-8000-000000000101";
export const DESIGN_OWNER_ID = "00000000-0000-4000-8000-000000000102";
export const DESIGN_VERSION_ID = "00000000-0000-4000-8000-000000000103";
export const DESIGN_ALTERNATIVE_ID = "00000000-0000-4000-8000-000000000104";
export const SECOND_DESIGN_ALTERNATIVE_ID = "00000000-0000-4000-8000-000000000105";
export const DESIGN_DIFF_ID = "00000000-0000-4000-8000-000000000106";
export const DESIGN_GATE_ID = "00000000-0000-4000-8000-000000000107";
export const DESIGN_CREATED_AT = "2026-08-21T10:00:00Z";

const REQUIREMENT_ID = "00000000-0000-4000-8000-000000000110";
const STORY_ID = "00000000-0000-4000-8000-000000000111";
const CRITERION_ID = "00000000-0000-4000-8000-000000000112";
const TWIN_ID = "00000000-0000-4000-8000-000000000113";

const TWIN_REFERENCE = {
  twin_id: TWIN_ID,
  version_number: 1,
  content_hash: "e".repeat(64),
  name: "Receptionist Twin",
};

export const BASE_DESIGN_PACKAGE: DesignPackagePayload = {
  schema_version: 1,
  project_id: DESIGN_PROJECT_ID,
  grounding: {
    requirements_reference: {
      kind: "REQUIREMENTS_SPECIFICATION",
      artifact_id: "00000000-0000-4000-8000-000000000120",
      version_number: 1,
      content_hash: "a".repeat(64),
    },
    agent_team_reference: {
      kind: "AGENT_TEAM",
      artifact_id: "00000000-0000-4000-8000-000000000121",
      version_number: 1,
      content_hash: "b".repeat(64),
    },
    user_modeling_reference: {
      kind: "USER_MODELING",
      artifact_id: "00000000-0000-4000-8000-000000000122",
      version_number: 1,
      content_hash: "c".repeat(64),
    },
    catalog: {
      version: 1,
      content_hash: "d".repeat(64),
    },
    requirement_ids: [REQUIREMENT_ID],
    user_story_ids: [STORY_ID],
    acceptance_criterion_ids: [CRITERION_ID],
    user_twin_references: [TWIN_REFERENCE],
  },
  alternatives: [
    {
      id: DESIGN_ALTERNATIVE_ID,
      code: "DES-001",
      approach: "GUIDED_WORKFLOW",
      title: "Guided reservation flow",
      summary: "Guide the receptionist through one decision at a time.",
      rationale: "Reduce avoidable cognitive load while keeping recovery visible.",
      requirement_ids: [REQUIREMENT_ID],
      user_story_ids: [STORY_ID],
      acceptance_criterion_ids: [CRITERION_ID],
      user_twin_references: [TWIN_REFERENCE],
      workflows: [
        {
          id: "00000000-0000-4000-8000-000000000130",
          code: "FLOW-001",
          title: "Create a reservation",
          steps: ["Review availability.", "Enter guest details.", "Confirm the reservation."],
          requirement_ids: [REQUIREMENT_ID],
          user_story_ids: [STORY_ID],
        },
      ],
      information_architecture: ["Availability", "Reservation", "Confirmation"],
      accessibility_considerations: ["Keep persistent labels for every control."],
      security_considerations: ["Minimize guest data shown in summaries."],
      advantages: ["The current step remains explicit."],
      trade_offs: ["Frequent users may need additional navigation."],
      assumptions: [],
      open_questions: ["Should an expert shortcut be added later?"],
    },
    {
      id: SECOND_DESIGN_ALTERNATIVE_ID,
      code: "DES-002",
      approach: "DASHBOARD_FIRST",
      title: "Reservation operations dashboard",
      summary: "Keep operational status and frequent actions visible together.",
      rationale: "Support rapid orientation across active reservation work.",
      requirement_ids: [REQUIREMENT_ID],
      user_story_ids: [STORY_ID],
      acceptance_criterion_ids: [CRITERION_ID],
      user_twin_references: [TWIN_REFERENCE],
      workflows: [
        {
          id: "00000000-0000-4000-8000-000000000131",
          code: "FLOW-002",
          title: "Review and create reservations",
          steps: ["Review status.", "Open the action panel.", "Confirm the update."],
          requirement_ids: [REQUIREMENT_ID],
          user_story_ids: [STORY_ID],
        },
      ],
      information_architecture: ["Overview", "Work queue", "Action panel"],
      accessibility_considerations: ["Expose status changes programmatically."],
      security_considerations: ["Reveal detailed guest data deliberately."],
      advantages: ["Frequent actions remain close to status information."],
      trade_offs: ["Higher information density may increase scanning effort."],
      assumptions: [],
      open_questions: [],
    },
  ],
  critiques: [
    {
      id: "00000000-0000-4000-8000-000000000140",
      code: "CRQ-001",
      kind: "SYNTHETIC_USER_TWIN",
      design_alternative_id: DESIGN_ALTERNATIVE_ID,
      user_twin_reference: TWIN_REFERENCE,
      strengths: ["The primary task remains visible."],
      concerns: ["The guided flow may slow down frequent users."],
      unmet_needs: [],
      accessibility_observations: ["Persistent labels support recovery."],
      trust_concerns: ["Real-user trust has not been established."],
      questions: ["Does this sequence match reception work during peak periods?"],
      suggested_changes: ["Validate optional keyboard shortcuts."],
      provenance: [
        {
          source_kind: "MODEL_OUTPUT",
          source_id: "fake-deterministic-design",
          source_version: 1,
          content_hash: null,
          locator: "alternatives.DES-001",
          summary: "Deterministic synthetic critique.",
        },
      ],
      confidence: 0.6,
      epistemic_status: "MODEL_INFERRED",
      human_validation: "REQUIRED",
      rationale: "This is simulated feedback derived from the approved User Twin profile.",
    },
  ],
  recommended_alternative_id: DESIGN_ALTERNATIVE_ID,
  owner_selected_alternative_id: null,
  prototype: null,
  concerns: [
    {
      id: "00000000-0000-4000-8000-000000000150",
      code: "DRK-001",
      summary: "Experienced users may find the guided flow slower.",
      mitigation: "Evaluate keyboard-efficient shortcuts after owner selection.",
      requirement_ids: [REQUIREMENT_ID],
      design_alternative_ids: [DESIGN_ALTERNATIVE_ID],
    },
  ],
  open_questions: ["Which direction should proceed to architecture planning?"],
};

export const SELECTED_DESIGN_PACKAGE = buildSelectedDesignPackage(
  BASE_DESIGN_PACKAGE,
  DESIGN_ALTERNATIVE_ID,
);

export const UNSELECTED_DESIGN_VERSION: DesignPackageVersionPayload = {
  id: DESIGN_VERSION_ID,
  project_id: DESIGN_PROJECT_ID,
  version_number: 1,
  based_on_version_number: null,
  content_hash: "f".repeat(64),
  package: BASE_DESIGN_PACKAGE,
  created_by_user_id: DESIGN_OWNER_ID,
  created_at: DESIGN_CREATED_AT,
  ready_for_gate: false,
};

export const SELECTED_DESIGN_VERSION: DesignPackageVersionPayload = {
  ...UNSELECTED_DESIGN_VERSION,
  id: "00000000-0000-4000-8000-000000000160",
  version_number: 2,
  based_on_version_number: 1,
  content_hash: "1".repeat(64),
  package: SELECTED_DESIGN_PACKAGE,
  ready_for_gate: true,
};

export const PROPOSED_DESIGN_DIFF: DesignPackageDiffPayload = {
  id: DESIGN_DIFF_ID,
  project_id: DESIGN_PROJECT_ID,
  owner_user_id: DESIGN_OWNER_ID,
  base_version_id: DESIGN_VERSION_ID,
  base_version_number: 1,
  base_content_hash: UNSELECTED_DESIGN_VERSION.content_hash,
  proposed_package: SELECTED_DESIGN_PACKAGE,
  proposal_hash: "2".repeat(64),
  changes: [
    {
      kind: "REPLACE",
      artifact_kind: "SELECTION",
      artifact_id: DESIGN_ALTERNATIVE_ID,
      before: null,
      after: {
        owner_selected_alternative_id: DESIGN_ALTERNATIVE_ID,
      },
    },
    {
      kind: "ADD",
      artifact_kind: "PROTOTYPE",
      artifact_id: SELECTED_DESIGN_PACKAGE.prototype?.id ?? DESIGN_ALTERNATIVE_ID,
      before: null,
      after:
        SELECTED_DESIGN_PACKAGE.prototype === null
          ? null
          : { ...SELECTED_DESIGN_PACKAGE.prototype },
    },
  ],
  status: "PROPOSED",
  created_at: DESIGN_CREATED_AT,
  decided_by_user_id: null,
  decided_at: null,
  decision_reason: null,
  applied_version_id: null,
  content_hash: "3".repeat(64),
};

export const PENDING_DESIGN_GATE: HumanGatePayload = {
  id: DESIGN_GATE_ID,
  project_id: DESIGN_PROJECT_ID,
  owner_user_id: DESIGN_OWNER_ID,
  gate_type: "DESIGN",
  artifact: {
    project_id: DESIGN_PROJECT_ID,
    gate_type: "DESIGN",
    artifact_id: SELECTED_DESIGN_VERSION.id,
    version: SELECTED_DESIGN_VERSION.version_number,
    content_hash: SELECTED_DESIGN_VERSION.content_hash,
  },
  iteration: 1,
  max_iterations: 3,
  status: "PENDING_APPROVAL",
  created_at: DESIGN_CREATED_AT,
  updated_at: DESIGN_CREATED_AT,
  event_sequence: 1,
  resume_status: null,
};
